common/src/stack/report-system/command/report/system/tests/test_sles_patch.py

```diff
@@ -0,0 +1,124 @@
+		pytest.skip('No SLES pallet found - skipping patch check')
+	sles_flavors = os.listdir(base_dir)
+	assert sles_flavors
+
+	# Find out where is stack-sles*images*.rpm file(s)
+	result = host.run('find /export/stack/pallets/stacki/ -name "*stack-sles-*.rpm"')
+	RPM = result.stdout.splitlines()
+	assert RPM
+
+	# If the stack-sles*images*.rpm installed?
+	result = host.run('rpm -qa | grep stack | grep images')
+	assert result
+
+	# Test every sles flavor found
+	for sles_flavor in sles_flavors:
+		# Make sure installed patches match what's under /opt/stack/pallet-patches
+		patch_dir = '/opt/stack/pallet-patches'
+		patch_dir_files = []
+		found_source = False
+		for (dir_path, dir_names, filenames) in os.walk(patch_dir):
+			# Only want particular SLES version
+			if sles_flavor in dir_path:
+				file_to_check = None
+				if '15sp' in sles_flavor:
+					file_to_check = 'CHECKSUMS'
+				else:
+					file_to_check = 'content'
+				patch_dir_files += [os.path.join(dir_path, file) for file in filenames if file == file_to_check]
+
+		# We should have non-empty list here
+		assert patch_dir_files
+
+		# Find img file from patch directory in SLES pallet
+		result = host.run(f'grep .img {patch_dir_files[0]}')
+
+		# Last item is the partial image path
+		part_img_file = result.stdout.split()[-1]
+
+		# Find full path to /export/stack/pallets/SLES/?
+		result = host.run(f'probepal {base_dir}')
+		assert result.rc == 0
+		palinfo = json.loads(result.stdout)
+
+        	# It shouldn't be empty
+		assert palinfo[base_dir]
+		
+		sles_pallet_root = None 
+		for pallet in palinfo[base_dir]:
+			# Grab the right sles sp level
+			if pallet['version'] == sles_flavor:
+				assert host.file(f"{pallet['pallet_root']}").is_directory
+				sles_pallet_root = pallet['pallet_root']
+				break
+
+		# .img file should exist in sles pallet directory in relative path
+		assert os.path.exists(sles_pallet_root + '/' + part_img_file)
+
+		# Verify all the stack-sles-* rpm packages 
+		for rpm in RPM:
+			# Make sure we found pm file
+			assert '.rpm' in rpm
+
+			result = host.run(f'rpm -qp --dump {rpm}')	
+			# We don't want file under /opt/stack/images to be included in this list
+			patch_files = [line.split() for line in result.stdout.splitlines() if sles_flavor in line and '/images/' not in line]
+
+			matched_list = []
+			# Verify file(s) from images RPM actually exist
+			for this_list in patch_files:
+				file_to_check = this_list[0]
+				assert os.path.exists(file_to_check) 
+				hash_value = this_list[3]
+
+				# Grab hash for this file in patch directory and SLES pallet directory and compare against hash from img file. They should match
+				# Need to build the equivalent path first
+				temp_path_list = file_to_check.split('add-stacki-squashfs')	
+				temp_path = sles_pallet_root + temp_path_list[1]
+				if get_hash(file_to_check,sles_flavor) == hash_value and get_hash(temp_path, sles_flavor) == hash_value:
+					matched_list.append(file_to_check)
```

Comment:
Shouldn't it be an failed assert if the hashed *don't* match? Right now they just don't end up on the matched_list, but I don't see any logic further on checking that all the RPMs that should have matched ended up having the correct hashes.

Reply:
Bill wanted to check all found SLES types and not abort on the first mismatch. So the found_source is declared/redefined for each SLES types. I check all found SLES types and print out which stacki is serving the OS image. If we don't find any stacki that is serving the found SLES type then raise an alert. I tested on FE with SLES11,12,15 BE and 2 versions of stacki installed.

Reply:
K. That makes sense. I think it would be safer to build up a list of which files didn't match and then check that the list is blank once you have gone through all the files. That way the logic will work even if there isn't 4 files. You could also output which file is corrupted or for whatever reason didn't match, which would be useful info to know.